File: UIViewPropertyAnimator/UIViewPropertyAnimatorObjCSample/UIViewPropertyAnimatorObjCSample/PropertyAnimatorViewController.m
//
//  ViewController.m
//  UIViewPropertyAnimatorObjCSample
//
//  Created by nikolay.andonov on 10/3/16.
//  Copyright © 2016 nikolay.andonov. All rights reserved.
//

#import "PropertyAnimatorViewController.h"

@interface PropertyAnimatorViewController ()

@property (weak, nonatomic) IBOutlet UIView *objectView;

@property (strong, nonatomic) UIViewPropertyAnimator *animator;
@property (assign, nonatomic) CGPoint targetLocation;

@end

@implementation PropertyAnimatorViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.objectView.backgroundColor = [self colorAtLocation:self.objectView.center];
    self.view.userInteractionEnabled = YES;
    self.targetLocation = self.objectView.center;
}

#pragma mark - Utilities

- (UIColor *)colorAtLocation:(CGPoint)location {
    
    CGFloat hue = (location.x / CGRectGetWidth([UIScreen mainScreen].bounds) + location.y / CGRectGetHeight([UIScreen mainScreen].bounds)) / 2;
    return [UIColor colorWithHue:hue saturation:0.78 brightness:0.75 alpha:1];
}

- (void)processTouches:(NSSet<UITouch *> *)touches {
    
    UITouch *touch = [touches anyObject];
    if (!touch) {
        return;
    }
    
    CGPoint loc = [touch locationInView:self.view];
    if (CGPointEqualToPoint(loc, self.targetLocation)) {
        return;
    }
    
    [self animateToLocation:loc];
}

- (void)animateToLocation:(CGPoint)location {
    
    NSTimeInterval duration = 0.6;
    id<UITimingCurveProvider> timing = [[UISpringTimingParameters alloc] initWithDampingRatio:0.5];
    self.animator = [[UIViewPropertyAnimator alloc] initWithDuration:duration timingParameters:timing];
    
    __weak typeof(self) weakSelf = self;
    self.animator = [[UIViewPropertyAnimator alloc] initWithDuration:duration dampingRatio:0.5 animations:^{
        weakSelf.objectView.center = location;
        weakSelf.objectView.backgroundColor = [weakSelf colorAtLocation:location];
    }];
    [self.animator startAnimation];
    
    self.targetLocation = location;
}

#pragma mark - Touch handlers

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    
    [self processTouches:touches];
}

- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    
    [self processTouches:touches];
}





@end

File: UIViewPropertyAnimator/UIViewPropertyAnimatorSwiftSample/UIViewPropertyAnimatorSwiftSample/PropertyAnimatorViewController.swift
//
//  ViewController.swift
//  UIViewPropertyAnimatorSwiftSample
//
//  Created by nikolay.andonov on 10/3/16.
//  Copyright © 2016 nikolay.andonov. All rights reserved.
//

import UIKit

class PropertyAnimatorViewController: UIViewController {

    private var targetLocation: CGPoint!
    private var animator : UIViewPropertyAnimator?
    
    @IBOutlet weak var objectView: UIView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        objectView.backgroundColor = colorAt(location: objectView.center)
        targetLocation = objectView.center
    }

    // MARK: - Utilities
    
    private func colorAt(location: CGPoint) -> UIColor {
        
        let hue: CGFloat = (location.x / UIScreen.main.bounds.width + location.y / UIScreen.main.bounds.height) / 2
        return UIColor(hue: hue, saturation: 0.78, brightness: 0.75, alpha: 1)
    }
    
    private func processTouches(_ touches: Set<UITouch>) {
        
        guard let touch = touches.first else {return}
        let loc = touch.location(in: view)
        
        if loc == targetLocation {
            return
        }
        
        animateTo(location: loc)
    }
    
    private func animateTo(location: CGPoint) {
        
        let duration: TimeInterval = 0.6
        let timing: UITimingCurveProvider = UISpringTimingParameters(dampingRatio: 0.5)
        
        animator = UIViewPropertyAnimator(
            duration: duration,
            timingParameters: timing)
        
        weak var weakSelf = self
        animator = UIViewPropertyAnimator.init(duration: duration, dampingRatio: 0.5, animations: { 
            
            weakSelf?.objectView.center = location
            weakSelf?.objectView.backgroundColor = self.colorAt(location: location)
        })
        
        animator?.startAnimation()
        targetLocation = location
    }
    
    
    // MARK: - Touch handlers
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        
        processTouches(touches)
    }
    
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        
        processTouches(touches)
    }

}

